src/client.ts

```diff
@@ -852,6 +938,209 @@ export class StreamClient<T extends QueryValue = any> {
+  /** Whether or not another page can be fetched by the client */
+  #isDone?: boolean;
+
+  /**
+   *
+   * @param query - A lambda that returns a promise for a {@link StreamToken}
+   * @param clientConfiguration - The {@link ChangeFeedClientConfiguration} to apply
+   * @example
+   * ```typescript
+   *  const changeFeed = client.changeFeed(streamToken);
+   * ```
+   */
+  constructor(
+    token: StreamToken | (() => Promise<StreamToken>),
+    clientConfiguration: ChangeFeedClientConfiguration,
+  ) {
+    if (token instanceof StreamToken) {
+      this.#query = () => Promise.resolve(token);
+    } else {
+      this.#query = token;
+    }
+
+    this.#clientConfiguration = clientConfiguration;
+    this.#lastCursor = clientConfiguration.cursor;
+
+    this.#validateConfiguration();
+  }
+
+  #getHeaders(): Record<string, string> {
+    return {
+      Authorization: `Bearer ${this.#clientConfiguration.secret}`,
+      "x-format": "tagged",
+      "x-driver-env": ChangeFeedClient.#driverEnvHeader,
+    };
+  }
+
+  async #nextPageHttpRequest() {
+    // If we never resolved the stream token, do it now since we need it here when
+    // building the payload
+    if (!this.#streamToken) {
+      this.#streamToken = await this.#resolveStreamToken(this.#query);
+    }
+
+    const headers = this.#getHeaders();
+
+    const req: HTTPRequest<ChangeFeedRequest> = {
+      headers,
+      client_timeout_ms: this.#clientConfiguration.query_timeout_ms,
+      data: {
+        token: this.#streamToken.token,
+      },
+      method: "POST",
+      path: FaunaAPIPaths.CHANGE_FEED,
+    };
+
+    // Set the page size if it is available
+    if (this.#clientConfiguration.page_size) {
+      req.data.page_size = this.#clientConfiguration.page_size;
+    }
+
+    // If we have a cursor, use that. Otherwise, use the start_ts if available.
+    // When the config is validated, if both are set, an error is thrown.
+    if (this.#lastCursor) {
+      req.data.cursor = this.#lastCursor;
+    } else if (this.#clientConfiguration.start_ts) {
+      req.data.start_ts = this.#clientConfiguration.start_ts;
+    }
+
+    return req;
+  }
+
+  async *[Symbol.asyncIterator](): AsyncGenerator<ChangeFeedPage<T>> {
+    while (!this.#isDone) {
+      yield await this.nextPage();
+    }
+  }
+
+  /**
+   * Fetches the next page of the change feed. If there are no more pages to
+   * fetch, this method will throw a {@link ClientError}.
+   */
+  async nextPage(): Promise<ChangeFeedPage<T>> {
+    if (this.#isDone) {
+      throw new ClientError("The change feed has no more pages to fetch.");
+    }
+
+    const { httpClient } = this.#clientConfiguration;
+
+    const request = await this.#nextPageHttpRequest();
+    const response = await withRetries(() => httpClient.request(request), {
+      maxAttempts: this.#clientConfiguration.max_attempts,
+      maxBackoff: this.#clientConfiguration.max_backoff,
+      shouldRetry: (error) => error instanceof ThrottlingError,
```

Comment:
I don't think so. The only thing we use built-in retries for is 429 errors. Or at least, the consistent thing to do with the current query API is to retry only on 429 errors.

Looking beyond the scope of this PR, I still don't think we should, but probably worthwhile checking with the Core team if 503's are handled well enough to be considered retryable. You might think so, but historically, queries that result in 503's tend to do so consistently (i.e. not a good idea to retry).

Reply:
that's highly unusual - most web services retry on 500 because, while outages are often clustered and some requests will be poison pills, many 5xx are intermittent or only affect a subset of the fleet.

Reply:
Let's do a follow-up for this one.